nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -558,52 +654,374 @@ async fn ssc_send_snapshot_request(
+    sagactx: NexusActionContext,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., authz_disk, db_disk) =
+        LookupPath::new(&opctx, &osagactx.datastore())
+            .disk_id(params.disk_id)
+            .fetch_for(authz::Action::Modify)
+            .await
+            .map_err(ActionError::action_failed)?;
+
+    info!(
+        log,
+        "attaching disk {} to pantry with id {:?}",
+        params.disk_id,
+        PANTRY_SENTINEL_ID.to_string(),
+    );
+
+    osagactx
+        .datastore()
+        .disk_update_runtime(
+            &opctx,
+            &authz_disk,
+            &db_disk.runtime().attach(*PANTRY_SENTINEL_ID),
+        )
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    Ok(())
+}
 
-            info!(
-                log,
-                "disk volume construction request {}",
-                disk_volume.data()
-            );
-
-            let disk_volume_construction_request: VolumeConstructionRequest =
-                serde_json::from_str(&disk_volume.data()).map_err(|e| {
-                    ActionError::action_failed(Error::internal_error(&format!(
-                        "failed to deserialize disk {} volume data: {}",
-                        disk.id(),
+async fn ssc_attach_disk_to_pantry_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let (.., authz_disk, db_disk) =
+        LookupPath::new(&opctx, &osagactx.datastore())
+            .disk_id(params.disk_id)
+            .fetch_for(authz::Action::Modify)
+            .await
+            .map_err(ActionError::action_failed)?;
+
+    match db_disk.runtime().attach_instance_id {
+        Some(attach_instance_id) => {
+            if attach_instance_id == *PANTRY_SENTINEL_ID {
+                info!(
+                    log,
+                    "undo: detaching disk {} from the pantry", params.disk_id
+                );
+
+                osagactx
+                    .datastore()
+                    .disk_update_runtime(
+                        &opctx,
+                        &authz_disk,
+                        &db_disk.runtime().detach(),
+                    )
+                    .await
+                    .map_err(ActionError::action_failed)?;
+            } else {
+                info!(
+                    log,
+                    "undo: disk {} not attached to pantry, it's attached to {}!",
+                    params.disk_id,
+                    attach_instance_id,
+                );
+
+                // XXX what to do here?
```

Comment:
I'm really not sure what to do here. I'm also not sure if this situation can ever be reached!

The forward running of this saga should have "attached" the disk to the pantry. If that failed, then this undo node wouldn't be executed during the unwind. If that node succeeded, then the disk is "attached" to the pantry. The only way we would end up in this branch is if something else then took the already attached disk and attached it somewhere else, which seems like a violation of some contract somewhere :)

Reply:
Maybe we log a message and return error?  If we get here, then we need higher level help to unwind things.

Reply:
I added a message, but since we are in the undo path, this could be an expected situation.